### PR TITLE
add2app test

### LINKED ISSUE #24117: Create an Add-to-App devicelab test for iOS
Recent changes to support Add2App broke some existing functionality that we test internally.

We should have a test in device lab to cover these scenarios.

/cc @cbracken @chinmaygarde @matthew-carroll @mit-mit @amirh (maybe something for maps/webview would make sense here?)

### PR DESCRIPTION
**Pull request by dnfield**
This will fix https://github.com/flutter/flutter/issues/24117

Runs some earl grey tests on an iOS simulator on Cirrus.

### CODE REVIEW

**Comment by dnfield** [comment]
This is failing as expected.  Once https://github.com/flutter/engine/pull/7610 lands, we should be able to fix the failing test and make this work.  Once that happens, we can just delete the existing ios_add2app repo. I was under the impression it would be important to have that separately from this because I thought we'd have to run certain commands in it that we don't actually have to to run these tests.

**Comment by dnfield** [comment]
@cbracken sorry about the size of this PR, but this is basically all of the content from github.com/flutter/ios_add2app.  It figured out a way to test it without Flutter yelling at me for living in the Flutter repo.

**Comment by dnfield** [comment]
cc @xster for ios review, @gspencergoog for cirrus review

File: .cirrus.yml [hunk 1]
```diff
@@ -165,6 +165,27 @@ task:
...
+    - analyze
+  env:
+    # Name the SDK directory to include a space so that we constantly
+    # test path names with spaces in them.
+    CIRRUS_WORKING_DIR: "/tmp/flutter add2app"
+    SHARD: add2app_test
+    COCOAPODS_DISABLE_STATS: true
+  osx_instance:
+    # note: the high sierra image doesn't seem to include a simulator.
+    image: mojave-xcode-10.1
```
Comment (gspencergoog) [inline .cirrus.yml]: Does it work to just have this `osx_instance` tag inside of one of the matrix nodes below?  I assume not, but you never know..  That way you could make sure this is being executed with most of the other stuff set up the same way (like artifact/package/pub caching, etc.).

Comment (dnfield) [reply .cirrus.yml]: I've bumped all our macOS to mojave and we'll see if it works.  Unfortunately right now this is blocked on a issue with analysis for the analyzer package.

**Comment by dnfield** [comment]
@xster I've implemented Greg's suggestions and gotten that squared away.  The Objective C app is ready for review.

**Comment by gspencergoog** [comment]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

For the Cirrus config, at least.

File: dev/integration_tests/android_views/ios/Podfile [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# Uncomment this line to define a global platform for your project
```
Comment (xster) [inline dev/integration_tests/android_views/ios/Podfile]: Is this file intentional?

File: dev/integration_tests/ios_add2app/flutterapp/.android.bak/Flutter/build.gradle [hunk 3]
```diff
@@ -0,0 +1,49 @@
+// Generated file. Do not edit.
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/.android.bak/Flutter/build.gradle]: How come all these files are in a .bak directory? Is that intentional?

File: dev/integration_tests/ios_add2app/Podfile [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+if !File.exist?(engine_dir)
+  # Copy the debug engine to have something to link against if the xcode backend script has not run yet.
+  debug_framework_dir = File.join(flutter_root(flutter_application_path), 'bin', 'cache', 'artifacts', 'engine', 'ios')
+  FileUtils.mkdir_p(engine_dir)
+  FileUtils.cp_r(File.join(debug_framework_dir, 'Flutter.framework'), engine_dir)
+  FileUtils.cp(File.join(debug_framework_dir, 'Flutter.podspec'), engine_dir)
+end
+
+target 'ios_add2app' do
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/Podfile]: remove extra space

File: dev/integration_tests/ios_add2app/README.md [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# iOS Add2App Test
+
+This application demonstrates some basic functionality for Add2App,
+along with a native iOS ViewController as a baseline and to demonstrate
+interaction.
+
+The following functionality is currently implemented:
+
+1. A regular iOS view controller, similar to the default `flutter create` template.
```
Comment (xster) [inline dev/integration_tests/ios_add2app/README.md]: A regular fullscreen instance of a FlutterViewController? Regular iOS view controller sounds non-flutter related.

File: dev/integration_tests/ios_add2app/README.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+1. A FlutterViewController subclass that takes over full screen. Demos showing this both from a cold/fresh engine state and a warm engine state.
```
Comment (xster) [inline dev/integration_tests/ios_add2app/README.md]: Does the subclass do anything? Makes one ask what's the difference between this one and 1.

File: dev/integration_tests/ios_add2app/README.md [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+1. A demo of pushing a FlutterViewController on as a child view.
+1. A demo of showing both the native and the Flutter views using a platform channel to to interact with each other.
+1. A demo of showing two FlutterViewControllers simultaneously.
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/README.md]: Could be useful adding FlutterViewController pushing a native view controller pushing a FlutterViewController and the second FlutterViewController doesn't affect the first one. Feel free to leave a bug and do later.

File: dev/integration_tests/ios_add2app/build_and_test.sh [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -e
+
+cd "$(dirname "$0")"
+
+pushd flutterapp
+../../../../bin/flutter build ios --debug --no-codesign -v
```
Comment (xster) [inline dev/integration_tests/ios_add2app/build_and_test.sh]: is -v necessary? Might make logs harder to parse.

File: dev/integration_tests/ios_add2app/flutterapp/lib/main.dart [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+import 'marquee.dart';
+
+/// Route names. (See [main] for more details.)
+///
+/// The route names must match those sent by the platform-specific component.
+const String greenMarqueeRouteName = 'marquee_green';
+const String purpleMarqueeRouteName = 'marquee_purple';
+const String fullscreenRouteName = 'full';
+const String hybridRouteName = 'hybrid';
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/lib/main.dart]: Add a comment for what this channel is for?

File: dev/integration_tests/ios_add2app/flutterapp/lib/main.dart [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+  _kReloadChannel.setMessageHandler(run);
+  run(ui.window.defaultRouteName);
+}
+
+Future<String> run(String name) async {
+  // The platform-specific component will call [setInitialRoute] on the Flutter
+  // view (or view controller for iOS) to set [ui.window.defaultRouteName].
+  // We then dispatch based on the route names to show different Flutter
+  // widgets.
+  // Since we don't really care about navigating in this app, we're not using
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/lib/main.dart]: s/navigating/Flutter-side navigation

File: dev/integration_tests/ios_add2app/flutterapp/lib/main.dart [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+class MyHomePage extends StatefulWidget {
+  const MyHomePage({this.initialRoute});
+
+  @override
+  _MyHomePageState createState() => _MyHomePageState();
+
+  final String initialRoute;
+
+  /// Whether we should display the home page in fullscreen mode.
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/lib/main.dart]: Thanks for the docs

File: dev/integration_tests/ios_add2app/flutterapp/lib/main.dart [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+const String purpleMarqueeRouteName = 'marquee_purple';
+const String fullscreenRouteName = 'full';
+const String hybridRouteName = 'hybrid';
+
+const String _kReloadChannelName = 'reload';
+const BasicMessageChannel<String> _kReloadChannel =
+    BasicMessageChannel<String>(_kReloadChannelName, StringCodec());
+
+void main() {
+  _kReloadChannel.setMessageHandler(run);
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/lib/main.dart]: From reading the README, not sure when this is supposed to happen. Add comment

File: dev/integration_tests/ios_add2app/flutterapp/lib/main.dart [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+        crossAxisAlignment: CrossAxisAlignment.start,
+        children: <Widget>[
+          Expanded(
+            child: Column(
+              mainAxisAlignment: MainAxisAlignment.center,
+              children: <Widget>[
+                Center(
+                  child: Text(
+                      '$buttonName tapped $_counter time${_counter == 1 ? '' : 's'}.',
+                      style: const TextStyle(fontSize: 17.0)),
```
Comment (xster) [inline dev/integration_tests/ios_add2app/flutterapp/lib/main.dart]: vertically match parenths of the same hierarchy

File: dev/integration_tests/ios_add2app/ios_add2app/MainViewController.h [hunk 14]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2018 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import <UIKit/UIKit.h>
+#import <Flutter/Flutter.h>
+
+@interface MainViewController : UIViewController
+
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/ios_add2app/MainViewController.h]: extra space

File: dev/integration_tests/ios_add2app/ios_add2app/NativeViewController.m [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+- (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  return [self initWithDelegate:nil];
+}
+
+-(instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
+   return [self initWithDelegate:nil];
+}
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/ios_add2app/NativeViewController.m]: extra space

File: dev/integration_tests/ios_add2app/ios_add2appTests/IntegrationTests.m [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+- (void)validateCountsPlatform:(NSString*)labelPrefix count:(int)platformCount {
+  NSString* platformCountStr =
+  [NSString stringWithFormat:@"%@ button tapped %d times.", labelPrefix, platformCount];
+
+  [[[EarlGrey selectElementWithMatcher:grey_accessibilityID(@"counter_on_iOS")]
+    assertWithMatcher:grey_text(platformCountStr)] assertWithMatcher:grey_sufficientlyVisible()];
+}
+
```
Comment (xster) [inline dev/integration_tests/ios_add2app/ios_add2appTests/IntegrationTests.m]: extra spaces

**Review by xster — APPROVED** [review]
LGTM

File: dev/integration_tests/ios_add2app/README.md (hunk 5 above)
Comment (dnfield) [reply dev/integration_tests/ios_add2app/README.md]: It is.  That one is just a regular iOS implementation for reference.  It's also used in one of the tests for platform interaction.

File: dev/integration_tests/ios_add2app/build_and_test.sh (hunk 8 above)
Comment (dnfield) [reply dev/integration_tests/ios_add2app/build_and_test.sh]: It's how we do almost all our tests...  we could miss some output on a failure if we don't do a -v (e.g. stacktraces aren't printed)

File: dev/integration_tests/ios_add2app/README.md (hunk 7 above)
Comment (dnfield) [reply dev/integration_tests/ios_add2app/README.md]: filed https://github.com/flutter/flutter/issues/28342

**Comment by dnfield** [comment]
@fkorotkov - it looks like the tests end up passing here, and on the `TEST SUCCEEDED!` things spin for 5 more minutes until it says agent has stopped responding.  This is on a Mac - it's the `add2app-macos` test.  Any idea what's going on here?  This worked yesterday...

**Comment by fkorotkov** [comment]
@dnfield logs are telling that the agent just stopped sending heartbeats. Weird that it occurred twice on this PR. Trying to reproduce...

**Comment by dnfield** [comment]
I have made some changes but nothing that should cause the test process to hang.  Are you able to see if any xcode related processes are still running after the success message?

**Comment by fkorotkov** [comment]
My task hasn't got to tests yet https://cirrus-ci.com/task/6590401151500288

**Comment by fkorotkov** [comment]
@dnfield I did VCNed into the VM that were running the task and witness a VM crash 😲

Took this screenshot a few seconds before.

screen shot 2019-02-22 at 4 50 03 pm

Not sure what's going on there. What is add2app doing differently than the other tasks?

**Comment by fkorotkov** [comment]
And yes. The task got to `** TEST SUCCEEDED **`. It seems that the VM crashed on a clean up? 🤔

**Comment by dnfield** [comment]
It runs xcodebuild and requires the simulator to be present to run some integration tests.

My last commit (where this started failing) was just some cleanup of documentation and unrelated files...

**Comment by fkorotkov** [comment]
Nothing has changed to the macOS VMs lately. The only suspicious thing might be that there are many concurrent VMs are running right now since it's working hours. I will investigate over the weekend and report back here.

**Comment by fkorotkov** [comment]
Haven't found anything on my own. Pinged Anka folks to help with directions.

**Comment by dnfield** [comment]
Fwiw I'm seeing a lot of timeouts and slowness this afternoon.  Hopefully will just pass

**Comment by fkorotkov** [comment]
@dnfield it seems it was a bug in the new nested virtualization feature which is enabled for mojave VMs but wasn't for High Sierra. Disabled it and now my task passed. Could you please re-run?

**Comment by dnfield** [comment]
Looks like it's working now, thanks
